### ISSUE TITLE
"Unknown options: extensions" error using flat config

### ISSUE THREAD

**Issue by crystalfp** [microsoft/vscode-eslint#1823]
With the move to ESlint 9.0 I enabled the use of flat config in VSC.
But the ESlint extension complains:
```
[Info  - 12:25:52 PM] ESLint server is starting.
[Info  - 12:25:52 PM] ESLint server running in node v18.18.2
[Info  - 12:25:52 PM] ESLint server is running.
[Info  - 12:25:57 PM] ESLint library loaded from: D:\Projects\STMng\node_modules\eslint\lib\api.js
[Error - 12:25:57 PM] An unexpected error occurred:
[Error - 12:25:57 PM] Error: Invalid Options:
- Unknown options: extensions
- 'extensions' has been removed.
    at processOptions (D:\Projects\STMng\node_modules\eslint\lib\eslint\eslint-helpers.js:805:15)
    at new ESLint (D:\Projects\STMng\node_modules\eslint\lib\eslint\eslint.js:592:34)
    at q (c:\Users\mvalle\.vscode\extensions\dbaeumer.vscode-eslint-2.4.4\server\out\eslintServer.js:1:18577)
    at E (c:\Users\mvalle\.vscode\extensions\dbaeumer.vscode-eslint-2.4.4\server\out\eslintServer.js:1:19075)
    at e.validate (c:\Users\mvalle\.vscode\extensions\dbaeumer.vscode-eslint-2.4.4\server\out\eslintServer.js:1:24921)
    at c:\Users\mvalle\.vscode\extensions\dbaeumer.vscode-eslint-2.4.4\server\out\eslintServer.js:1:220305
```
Apparently the extension works.

Version: 1.88.0
Commit: 5c3e652f63e798a5ac2f31ffd0d863669328dc4c
Date: 2024-04-03T13:26:18.741Z
Browser: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Code/1.88.0 Chrome/120.0.6099.291 Electron/28.2.8 Safari/537.36

**Comment by dbaeumer** [comment]
Does linting work for you correctly in the terminal. If so, can you please provide me with a GitHub repository I can clone with a minimal setup that demos the problem.

**Comment by pdsuwwz** [comment]
> Does linting work for you correctly in the terminal. If so, can you please provide me with a GitHub repository I can clone with a minimal setup that demos the problem.

Hi ~ I had the same problem...

I tried to make a mini reproduction: https://github.com/pdsuwwz/eslint-v9-mini-repo

It works on the terminal:

![image](https://github.com/microsoft/vscode-eslint/assets/19891724/43954bac-dcaa-41fc-bcc7-59d26fd24889)

but `vscode-eslint` still reports errors:

![image](https://github.com/microsoft/vscode-eslint/assets/19891724/a3f758a3-9f8a-4e16-929f-c7761b2ceb31)

**Comment by crystalfp** [comment]
Idem for me. It works from the terminal but not inside VSC.
Inside VSC some problem are reported and some not, for example missing semicolons.

**Comment by dbaeumer** [comment]
Which version of ESLint extension are you using. Good flat config support requires 3.0.5 pre-release.

**Comment by crystalfp** [comment]
I have version 2.4.4. This explain everything. When will be 3.0.5 available?

**Comment by crystalfp** [comment]
Same problem here.

**Comment by asakurayoh1987** [comment]
> Which version of ESLint extension are you using. Good flat config support requires 3.0.5 pre-release.

Yes, but the error is still persists

**Comment by dbaeumer** [comment]
@crystalfp 3.0.5 pre-release is available in the market place.

**Comment by dbaeumer** [comment]
@pdsuwwz very strange. The repository you provided me with works for me without problems

![Image](https://github.com/microsoft/vscode-eslint/assets/1931590/5eb28294-f736-4064-b132-d6e5288a4d4d)

**Comment by dbaeumer** [comment]
@crystalfp @asakurayoh1987 can you please provide me with a GitHub repository I can clone with a minimal setup (like the one @pdsuwwz  provided) that demos what you are seeing.

**Comment by pdsuwwz** [comment]
It’s strange. I don't know what went wrong, is it possible that vscode itself or some other vscode plugin affected it

**Comment by crystalfp** [comment]
Tried to bisect extensions and the only one generating the error is ESlint. Also disabling all extensions but ESlint gives again the error.

**Comment by crystalfp** [comment]
Linting ignore rules. For example rule `"@typescript-eslint/semi": ["error", "always"],` is silently ignored.
For now I return to non-flat config inside VSC. Your extension is so valuable for development that I cannot live without. But I'm always available for testing the flat-config version.

**Comment by dbaeumer** [comment]
@pdsuwwz the only thing I can think of is some global things you have on your machine that I don't have.

@crystalfp a GitHub repository I can clone that demos what you are seeing is still appreciated.

**Comment by pdsuwwz** [comment]
I've caught the issue! It turns out that there was an extension configuration option in my `User/settings.json` file. I removed it and ESLint is back to normal now:

In file: `~/Library/Application\ Support/Code/User/settings.json`


```json
 "eslint.options": {
    "extensions": [ ".html", ".js", ".vue", ".jsx" ]
},
```

![image](https://github.com/microsoft/vscode-eslint/assets/19891724/6847938b-9381-44f1-8b79-908ada59ba8a)

@crystalfp  @asakurayoh1987  If you're having the same issue, check your User settings.json for the config I mentioned.

**Comment by crystalfp** [comment]
Thanks a lot!
I overlooked this because VSC shows `eslint.options` and `eslint.validate` as greyed out.
Maybe the error message could specify where this option resides, so the correction could be immediate.
Thanks again!
mario

**Comment by fjordskii** [comment]
@pdsuwwz amazing find! Worked for me as well 💪

**Comment by asakurayoh1987** [comment]
> I've caught the issue! It turns out that there was an extension configuration option in my `User/settings.json` file. I removed it and ESLint is back to normal now:
> 
> In file: `~/Library/Application\ Support/Code/User/settings.json`
> 
> ```json
>  "eslint.options": {
>     "extensions": [ ".html", ".js", ".vue", ".jsx" ]
> },
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/19891724/321838937-6847938b-9381-44f1-8b79-908ada59ba8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM5MzA4NTcsIm5iZiI6MTcxMzkzMDU1NywicGF0aCI6Ii8xOTg5MTcyNC8zMjE4Mzg5MzctNjg0NzkzOGItOTM4MS00NGYxLThiNzktOTA4YWRhNTliYThhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI0VDAzNDkxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjNmIzNjJmN2IyNWM1ZmIxOTljZDg5MmU3NGNiNWQ5YTk5NTM1ZGVkYTkwZDQ5MDQ4NmEyMGQ3NDdkZWM3NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lOOtQqO9rH83HW3IeFFW8DxoJWl5pZh9ZjFDQoCiRTU)
> 
> @crystalfp @asakurayoh1987 If you're having the same issue, check your User settings.json for the config I mentioned.

It works! thx !

**Comment by vorlov** [comment]
having same issue with v 9.1.1, check this config, no extensions at all by text search on a project

**Comment by pdsuwwz** [comment]
You can press `ctrl + p`, enter `> open user settings json`  to open the corresponding json file, and you may be able to find the corresponding configuration information from:

![image](https://github.com/microsoft/vscode-eslint/assets/19891724/4df1c7b7-6431-41bd-9b8c-32b36a8e1eba)

@rajvatsal @vorlov

**Comment by rajvatsal** [comment]
@pdsuwwz thanx I found settings.json. I didn't knew it was something specific to vscode or eslint. My settings.json doesn't have the `eslint.options` property. I am facing this error in neovim.

**Comment by artemshchirov** [comment]
My project and settings.json also doesn't have the eslint.options property. I am also always get this error when building Next.js project in VSCode:

`web:build:  ⨯ ESLint: Invalid Options: - Unknown options: useEslintrc, extensions - 'extensions' has been removed.`

I don't have useEslintrc anywhere in my project or in settings.json either.

**Comment by dbaeumer** [comment]
> web:build:  ⨯ ESLint: Invalid Options: - Unknown options: useEslintrc, extensions - 'extensions' has been removed.

Is this when running in the terminal. Then I guess that is some npm script, ... that passes these arguments to eslint.

**Comment by MuraliPrasanth2** [comment]
> @pdsuwwz thanx I found settings.json. I didn't knew it was something specific to vscode or eslint. My settings.json doesn't have the `eslint.options` property. I am facing this error in neovim.
> 
> 

Did you find any solution, I am also facing the same problem in neovim.

**Comment by rajvatsal** [comment]
@dbaeumer It turns out that the binary that I was using `eslint_d` doesn't support latest eslint version. I switched to `eslint-lsp` and it works fine.

**Comment by rajvatsal** [comment]
> > @pdsuwwz thanx I found settings.json. I didn't knew it was something specific to vscode or eslint. My settings.json doesn't have the `eslint.options` property. I am facing this error in neovim.
> 
> Did you find any solution, I am also facing the same problem in neovim.

I was using null_ls with eslint_d for diagnostics but when I switched to eslint-lsp all the errors stopped. There is an issue in eslint_d that it doesn't support the latest eslint version right now. So I prefer you use eslint-lsp

**Comment by LAST7** [comment]
> > > @pdsuwwz thanx I found settings.json. I didn't knew it was something specific to vscode or eslint. My settings.json doesn't have the `eslint.options` property. I am facing this error in neovim.
> > 
> > 
> > Did you find any solution, I am also facing the same problem in neovim.
> 
> I was using null_ls with eslint_d for diagnostics but when I switched to eslint-lsp all the errors stopped. There is an issue in eslint_d that it doesn't support the latest eslint version right now. So I prefer you use eslint-lsp

@rajvatsal Thanks dude, never thought of the problem of eslint_d not catching up to the latest version of eslint. I'm using `none-ls` and `none-ls-extra`, so I added these lines inside the `source` option:

```lua
            require("none-ls.diagnostics.eslint").with {
                -- only enable eslint if root has eslint.config.mjs
                condition = function(utils)
                    return utils.root_has_file "eslint.config.mjs" -- change file extension if you use something else
                end,
            },
```

Previously I was using `none-ls.diagnostics.eslint_d`, and now I change `eslint_d` into `eslint`, and everything's working fine.

**Comment by jabdinghoff** [comment]
> My project and settings.json also doesn't have the eslint.options property. I am also always get this error when building Next.js project in VSCode:
> 
> `web:build: ⨯ ESLint: Invalid Options: - Unknown options: useEslintrc, extensions - 'extensions' has been removed.`
> 
> I don't have useEslintrc anywhere in my project or in settings.json either.

same for me trying to build a vue project

**Comment by JREAM** [comment]
> My project and settings.json also doesn't have the eslint.options property. I am also always get this error when building Next.js project in VSCode:
> 
> `web:build: ⨯ ESLint: Invalid Options: - Unknown options: useEslintrc, extensions - 'extensions' has been removed.`
> 
> I don't have useEslintrc anywhere in my project or in settings.json either.

Downgrade from  `eslint` version `9.x.x` to `8.57.0` and try it again.

**Comment by tomkersten** [comment]
> @dbaeumer It turns out that the binary that I was using eslint_d doesn't support latest eslint version. I switched to eslint-lsp and it works fine.

This pointed me in the right direction. Thank you!

I had a global installation of `eslint_d`. 

Running `eslint` in the terminal didn't give an error because it had been upgraded, but `eslint_d` had not. As of `eslint_d: v14.0.3, bundled eslint: v9.9.0` I had no issues and did not need to switch to `eslint-lsp`.

**Comment by awixor** [comment]
> User/settings.json

I checked the `User/settings.json` file but it doesn't have anything related to eslint, and I still have the issue on `v9.10.0`

**Comment by dbaeumer** [comment]
@awixor anything in the workspace settings. If not can you provide me with a GitHub repository I can clone with a minimal setup that reproduces the problem.

**Comment by awixor** [comment]
@dbaeumer no, nothing in there. 
I tried in both Windows and Mac machines but still, it doesn't work.
The thing I did to avoid it is using `pnpm create @eslint/config@0.4.6`  then downgrade eslint to `v8.57.0`, because even with old `@eslint/create-config` versions it still installs latest version of eslint.

**Comment by dbaeumer** [comment]
Without a minimal repro very hard to say what is going on.

**Comment by akseyh** [comment]
> @dbaeumer no, nothing in there. I tried in both Windows and Mac machines but still, it doesn't work. The thing I did to avoid it is using `pnpm create @eslint/config@0.4.6` then downgrade eslint to `v8.57.0`, because even with old `@eslint/create-config` versions it still installs latest version of eslint.

Worked for me. Thanks!

**Comment by awixor** [comment]
@dbaeumer I just 
-Created a new CRA app and 
- Run this `pnpm create @eslint/config@latest`
- `pnpm run start`

I pushed the repo here: https://github.com/awixor/investigate-eslint-v9

Error:
![image](https://github.com/user-attachments/assets/318efaa4-1c76-417c-9dc7-68fe88c67507)

**Comment by dbaeumer** [comment]
@awixor this is not produced by the ESLint extension. I would suggest that you report this to the tool chain that reports that problem.

Trying to validate that with ESLint in the terminal doesn't work either. It produces

./node_modules/.bin/eslint src/setupTests.ts 
Warning: React version not specified in eslint-plugin-react settings. See https://github.com/jsx-eslint/eslint-plugin-react#configuration 

Error: Failed to load config "react-app" to extend from.
Referenced from: /workspaces/investigate-eslint-v9/package.json
    at configInvalidError (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2648:9)
    at ConfigArrayFactory._loadExtendedShareableConfig (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3279:23)
    at ConfigArrayFactory._loadExtends (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3156:25)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3095:25)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3040:20)
    at _normalizeObjectConfigData.next (<anonymous>)
    at ConfigArrayFactory.loadInDirectory (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2886:28)
    at CascadingConfigArrayFactory._loadConfigInAncestors (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3871:46)
    at CascadingConfigArrayFactory._loadConfigInAncestors (/usr/local/share/npm-global/lib/node_modules/eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3890:20)

**Comment by awixor** [comment]
@dbaeumer  I am pretty sure that file in not the cause of the issue.
Can you delete `src/setupTests.ts` and try again.

**Comment by dbaeumer** [comment]
@awixor I am not saying that `src/setupTests.ts` is the source of the problem. What I am saying is the following:

- the error you see is NOT generated by the ESLint extension. It must be some other tooling that calls eslint that results in this and I don't know which tool that is.
- even if I try to validate a file using eslint in the terminal it fails. So there is something wrong with the general eslint setup which is unrelated to the extension. In general the extension works if linting files in the terminal works as well.

**Comment by serg-mois-capital** [comment]
After migration from eslint 8 to 9 - faced the same error. 

 Error: Invalid Options:
- Unknown options: extensions
- 'extensions' has been removed.

**Comment by nachodd** [comment]
I had the same issue. The problem was in my settings.json file on VSCode.
To fix it, open the command pallet `cmd+shift+P` > `Preferences: Open User Settings (JSON)` and just comment the `extensions` key under `eslint.options`. 
<img width="309" alt="Image" src="https://github.com/user-attachments/assets/691868ff-a822-4d32-9e2e-74b1ce8539b0" />
Then restart ESLint Server: `cmd+shift+P` > `ESLint: Restart ESLint Server`

**Comment by movy** [comment]
Rather outdated edge-case, but if you're still maintaining legacy Vue2 projects, eslint 9 is not supported (and will never be obviously). Stay on 8.57.0 and everything works even with flat config.

**Comment by ollyskinner** [comment]
> Rather outdated edge-case, but if you're still maintaining legacy Vue2 projects, eslint 9 is not supported (and will never be obviously), you'll get `"Unknown options: extensions" error using flat config` error with eslint 9
> 
> Stay on 8.57.0 and everything works even with flat config.

Thanks, this solved for my case, maintaining a 5-year-old Vue2 project!
